### PR TITLE
DX: Event Handler Types

### PR DESCRIPTION
**Pull request by nnelgxorz**
# What is it?

- [x] Feature / enhancement
- [ ] Bug
- [ ] Docs / tests

# Description

Adds autocompletions for host:on, document:on, window:on and preventDefault: attributes.

# Use cases and why

- 1. Improves developer experience by exposing autocompletions.
- 2. DRYs up event handler code. Add events to the QwikEventMap and all types are automatically generated.

# Checklist:

- [x] My code follows the [developer guidelines of this project](../CONTRIBUTING.md)
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] Added new tests to cover the fix / functionality

### CODE REVIEW

File: packages/qwik/src/core/render/jsx/types/jsx-qwik-attributes.ts [hunk 1]
```diff
@@ -185,10 +192,31 @@ interface CSSProperties {
...
+  [key: `${'host'}on:${string}Qrl`]: QrlEvent<any> | QrlEvent<any>[];
+  [key: `${'window' | 'document'}:on${string}$`]: NativeEventHandler<any> | undefined;
+  [key: `${'window' | 'document'}:on${string}Qrl`]: QrlEvent<any> | QrlEvent<any>[] | undefined;
+}
+export type ComponentKnownEvents$ = {
+  [K in keyof QwikEventMap as `${'host' | 'window' | 'document'}:on${K}$`]?: NativeEventHandler<
+    QwikEventMap[K]
+  >;
+};
+export type ComponentKnownEventsQrl = {
```
Comment (manucorporat) [inline packages/qwik/src/core/render/jsx/types/jsx-qwik-attributes.ts:206]: actually, this is not longer needed in main

File: packages/qwik/src/core/render/jsx/types/jsx-qwik-attributes.ts [hunk 2]
```diff
@@ -185,10 +192,31 @@ interface CSSProperties {
  */
 export type JSXTagName = keyof HTMLElementTagNameMap | Omit<string, keyof HTMLElementTagNameMap>;
 
+export interface ComponentCustomEvents {
+  [key: `${'host'}on:${string}$`]: NativeEventHandler<any>;
+  [key: `${'host'}on:${string}Qrl`]: QrlEvent<any> | QrlEvent<any>[];
+  [key: `${'window' | 'document'}:on${string}$`]: NativeEventHandler<any> | undefined;
+  [key: `${'window' | 'document'}:on${string}Qrl`]: QrlEvent<any> | QrlEvent<any>[] | undefined;
```
Comment (manucorporat) [inline packages/qwik/src/core/render/jsx/types/jsx-qwik-attributes.ts:199]: Same here, not needed Qrl ending

File: packages/qwik/src/core/render/jsx/types/jsx-qwik-attributes.ts [hunk 3]
```diff
@@ -185,10 +192,31 @@ interface CSSProperties {
  */
 export type JSXTagName = keyof HTMLElementTagNameMap | Omit<string, keyof HTMLElementTagNameMap>;
 
+export interface ComponentCustomEvents {
+  [key: `${'host'}on:${string}$`]: NativeEventHandler<any>;
+  [key: `${'host'}on:${string}Qrl`]: QrlEvent<any> | QrlEvent<any>[];
```
Comment (manucorporat) [inline packages/qwik/src/core/render/jsx/types/jsx-qwik-attributes.ts:197]: Not needed!

File: packages/qwik/src/core/render/jsx/types/jsx-qwik-attributes.ts [hunk 4]
```diff
@@ -170,10 +181,6 @@ export interface QwikEvents {
 
   'document:onVisibilityChange$'?: (event: Event, el: Element) => any;
   'document:onVisibilityChangeQrl'?: QRL<(event: Event, el: Element) => any>;
```
Comment (manucorporat) [inline packages/qwik/src/core/render/jsx/types/jsx-qwik-attributes.ts:183]: We can remove too

**Comment by manucorporat** [comment]
Love this PR! lets get it done and merge!

File: packages/qwik/src/core/render/jsx/types/jsx-qwik-attributes.ts (hunk 4 above)
Comment (manucorporat) [reply packages/qwik/src/core/render/jsx/types/jsx-qwik-attributes.ts:183]: the Qrl variant

Comment (nnelgxorz) [reply packages/qwik/src/core/render/jsx/types/jsx-qwik-attributes.ts:183]: Cool. Also, document:onLoad$ and document:onScroll$ should already be generated from the new types. Should I delete those too?

**Comment by manucorporat** [comment]
Havent tested this change, but i am worried about the autocompletion of events, like this:
Screenshot 2022-07-26 at 10 44 45

is that still working?

**Comment by nnelgxorz** [comment]
Yep! It's still working for me. But now unknown/custom events will have type Event instead of any. :+1:

**Comment by nnelgxorz** [comment]
It's a trick React uses for the same problem:

https://stackoverflow.com/questions/52667959/what-is-the-purpose-of-bivariancehack-in-typescript-types/52668133#52668133

**Comment by manucorporat** [comment]
Amazing! going to merge another PR first, then this one!

**Pull request merged**